emulator/computation_profile.go

```diff
@@ -0,0 +1,205 @@
+/*
+ * Flow Emulator
+ *
+ * Copyright 2019 Dapper Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package emulator
+
+import (
+	"github.com/onflow/cadence/runtime/common"
+	"github.com/onflow/flow-emulator/types"
+	"github.com/onflow/flow-go/fvm/environment"
+	"github.com/onflow/flow-go/fvm/meter"
+)
+
+type ProcedureReport struct {
+	ID              string          `json:"ID"`
+	ComputationUsed uint64          `json:"computation"`
+	Intensities     map[string]uint `json:"intensities"`
+	MemoryEstimate  uint64          `json:"memory"`
+	Code            string          `json:"source"`
+	Arguments       []string        `json:"arguments"`
+}
+
+type ComputationProfile struct {
+	Scripts      map[string]ProcedureReport `json:"scripts"`
+	Transactions map[string]ProcedureReport `json:"transactions"`
+}
+
+func (cp *ComputationProfile) ReportScript(
+	scriptResult *types.ScriptResult,
+	code string,
+	arguments []string,
+	intensities meter.MeteredComputationIntensities,
+) {
+	scriptReport := ProcedureReport{
+		ID:              scriptResult.ScriptID.String(),
+		ComputationUsed: scriptResult.ComputationUsed,
+		Intensities:     transformIntensities(intensities),
+		MemoryEstimate:  scriptResult.MemoryEstimate,
+		Code:            code,
+		Arguments:       arguments,
+	}
+	cp.Scripts[scriptResult.ScriptID.String()] = scriptReport
+}
+
+func (cp *ComputationProfile) ReportTransaction(
+	txResult *types.TransactionResult,
+	code string,
+	arguments []string,
+	intensities meter.MeteredComputationIntensities,
+) {
+	txReport := ProcedureReport{
+		ID:              txResult.TransactionID.String(),
+		ComputationUsed: txResult.ComputationUsed,
+		Intensities:     transformIntensities(intensities),
+		MemoryEstimate:  txResult.MemoryEstimate,
+		Code:            code,
+		Arguments:       arguments,
+	}
+	cp.Transactions[txResult.TransactionID.String()] = txReport
+}
+
+// transformIntensities maps a numeric common.ComputationKind value
+// to a human-friendly string value.
+func transformIntensities(
```

Comment:
I wish flow-go already offered ComputationKind names.

Reply:
Unfortunately, I think this is not possible. The main problem is that `ComputationKind` is defined in Cadence repository and it uses the `string` cmd tool to generate the `String` representation for the values defined there. See: https://github.com/onflow/cadence/blob/master/runtime/common/compositekind.go#L27
Only the `ComputationKind` values defined in Cadence offer a human-readable name: https://github.com/onflow/cadence/blob/master/runtime/common/computationkind_string.go.